common/src/main/scala/net/psforever/objects/avatar/PlayerControl.scala

```diff
@@ -669,20 +702,194 @@ class PlayerControl(player : Player) extends Actor
     val zone = obj.Zone
     zone.AvatarEvents ! AvatarServiceMessage(player.Name, AvatarAction.SendResponse(Service.defaultPlayerGUID, ObjectDetachMessage(obj.GUID, item.GUID, Vector3.Zero, 0f)))
   }
+
+  /**
+    * Determine whether the current stamina value for this player requires a greater change in player states.
+    * Losing all stamina and not yet being fatigued deactivates implants.
+    * Having stamina of 20 points or greater and having previously been fatigued
+    * allows implants to operate once again.
+    * Initialization must be restarted manually for any implant that had not previously finished initializing.
+    */
+  def UpdateStamina() : Unit = {
+    val currentStamina = player.Stamina
+    if(currentStamina == 0 && !player.Fatigued) { // Only be fatigued once even if loses all stamina again
+      player.Fatigued = true
+      player.skipStaminaRegenForTurns += 4
+      player.Implants.indices.foreach { slot => // Disable all implants
+        ImplantActivation(slot, status = 0)
+        player.Zone.AvatarEvents ! AvatarServiceMessage(player.Name, AvatarAction.SendResponse(Service.defaultPlayerGUID, AvatarImplantMessage(player.GUID, ImplantAction.OutOfStamina, slot, 1)))
+      }
+    }
+    else if(currentStamina >= 20) {
+      val wasFatigued = player.Fatigued
+      player.Fatigued = false
+      if(wasFatigued) { //reactivate only if we were fatigued
+        player.Implants.indices.foreach { slot => // Re-enable all implants
+          player.Zone.AvatarEvents ! AvatarServiceMessage(player.Name, AvatarAction.SendResponse(Service.defaultPlayerGUID, AvatarImplantMessage(player.GUID, ImplantAction.OutOfStamina, slot, 0)))
+          if(!player.ImplantSlot(slot).Initialized) {
+            ImplantInitializationStart(slot)
+          }
+        }
+      }
+    }
+    player.Zone.AvatarEvents ! AvatarServiceMessage(player.Name, AvatarAction.PlanetsideAttributeToAll(player.GUID, 2, player.Stamina))
```

Comment:
Do all other players need to know about the stamina state of a particular avatar with `PlanetsideAttributeToAll`? I don't think there's anything that relies on you knowing how much stamina another player has

Reply:
It doesn't matter if it's `PlanetsideAttributeToAll` or `PlanetsideAttributeToSelf` because it's only going to one channel - `player.Name`.  Also, no `if` check at the receiving `WSA`.

Reply:
I did miss that it was sent to `player.Name`, but is it worth changing to `PlanetsideAttributeSelf` anyway to make the intent a bit clearer? Looking at it in the future might be confusing that it's sent to one player, but is using `PlanetsideAttributeToAll` implying that all players should get it

Reply:
I do have a "I'll never get to it at this rate" ticket for doing stuff to the events messaging.

`...Self` also has an `if` check on the receiving end which is made pointless by ability to send to that player's specific channel.  There is a use for `...Self` - if that specific player, by GUID, should receive the message, but the player has respawned so it no longer matters they they get the message; or, if you only know the GUID of the player, but didn't track down their name.  That hedges against latency and timers.